### PR TITLE
Modernise FormCheckoutRevision

### PR DESCRIPTION
**Pull request by RussKie**
Contributes to #6183

## Proposed changes

- Modernise FormCheckoutRevision

### Before

image

### After

image

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormCheckoutRevision.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ private FormCheckoutRevision()
         }
 
         public FormCheckoutRevision(GitUICommands commands)
-            : base(commands)
+            : base(commands, enablePositionRestore: true)
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCheckoutRevision.cs:22]: Will we encounter the same effects as with FormCommit?

**Review by mstv — COMMENTED** [review]
The layout was changed in a slightly different way in #11158, particularly the CheckBoxes.
The Label and the TextBox are not aligned (to the baseline of the text).

File: GitUI/CommandsDialogs/FormCheckoutRevision.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormCheckoutRevision.cs:22]: A lot of dialogs have this on.

Thinking about this now, this should _probably_ be off.

**Comment by RussKie** [comment]
> The Label and the TextBox are not aligned (to the baseline of the text).

I didn't notice this, I'll fix that.



> The layout was changed in a slightly different way in #11158, particularly the CheckBoxes.

The proposed layout aligns with many other dialogs, such as create/delete branch.
In the context of that dialog moving the checkboxes to the left column made the dialog visually balanced.
Unless there's a full-width control immediately preceding a checkbox control, we put checkboxes in the second column.

I will also be changing the layout of this control in the follow up work adding the same "reset options" as found on the "checkout local branch" dialog.

**Comment by cristianst85** [comment]
@RussKie, shouldn't this also be a modal dialog (i.e., no minimize and maximize buttons)?

**Comment by mstv** [comment]
> Shouldn't this also be a modal dialog with no minimize and maximize buttons? Like a `FixedDialog` type.

The maximize button is not needed but does not harm.
Though all (modal) dialogs shall have a minimize button. Why should we deny to minimize the entire app?

**Comment by RussKie** [comment]
I don't think minimise button will minimise the app. Something to check.

**Comment by mstv** [comment]
> I don't think minimise button will minimise the app. Something to check.

It does perfectly.

**Comment by cristianst85** [comment]
I was saying that because some modals don't have neither minimize nor maximize (e.g., Create tag, Reset branch) hence the inconsistency.

**Comment by RussKie** [comment]
Thank you for pointing this out. I'll fix it.

**Comment by RussKie** [comment]
> shouldn't this also be a modal dialog with no minimize and maximize buttons? Like a `FixedDialog` type.
>
> I was saying that because some (?) modals don't have neither minimize nor maximize (e.g., Create tag, Create new branch, Reset branch, Cherry pick commit) - most of them don't (I think) - hence the inconsistency in design with other dialogs.

Updated





> The Label and the TextBox are not aligned (to the baseline of the text).

It's not that simple due to the layout of the `CommitPickerSmallControl` control (which contains the textbox and the revision selector button). I've set the label's text alignmet to the v-center but to me it looks slightly worse.

File: GitUI/CommandsDialogs/FormCheckoutRevision.Designer.cs [hunk 2]
```diff
@@ -124,6 +125,8 @@ private void InitializeComponent()
             AutoScaleDimensions = new SizeF(96F, 96F);
             AutoScaleMode = AutoScaleMode.Dpi;
             ClientSize = new Size(481, 131);
+            MaximizeBox = false;
+            MinimizeBox = false;
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCheckoutRevision.Designer.cs:129]:
```suggestion
            MaximizeBox = false;
```

**Review by mstv — DISMISSED** [review]
> Though all (modal) dialogs shall have a minimize button. Why should we deny to minimize the entire app?

>> I don't think minimise button will minimise the app.

> It does perfectly.

**Comment by RussKie** [comment]
This is how many other dialogs are configured.

**Comment by mstv** [comment]
> This is how many other dialogs are configured.

So, those should get a minimize button, too.

**Comment by RussKie** [comment]
That's a separate unrelated work.

**Comment by RussKie** [comment]
> Again: Why should we deny to minimize the entire app?

This and other dialogs (such as `CreateBranch`, `DeleteBranch`, `RenameBranch`) are short-lived, so it's arguable that these dialogs must facilitate the "minimise the app" functionality.
I do feel that this discussion is completely tangential to this change.

**Comment by mstv** [comment]
> dialogs (such as CreateBranch, DeleteBranch, RenameBranch) are short-lived,

Branch names often need to contain a ticket number and / or a target SW version. So switching apps from these dialogs is not unusual.

> so it's arguable that these dialogs must facilitate the "minimise the app" functionality.

That many UX designers think like this, does not mean they should decide to disable a feature which is useful for some users - maybe only a few.
I do not see the cost of having these buttons in the title bar. They do not harm. If you do not want to use them, do not click them.
Basically, many dialogs (not this one) benefit from being resizable (always including maximization).

> I do feel that this discussion is completely tangential to this change.

The feature has arbitrarily been disabled here - overriding a sensible default...

**Comment by RussKie** [comment]
> > I do feel that this discussion is completely tangential to this change.
> 
> The feature has arbitrarily been disabled here - overriding a sensible default...

No, it's been disabled for consistency with other dialogs.
image

**Comment by mstv** [comment]
If this has been done / has not been considered for these many dialogs, it does not imply that this is the good UX.
This PR is a regression regarding this UX aspect (minimize option).
Why should we restrict users in how they can work with multiple apps? It always is a nuisance if an app cannot be minimized. The `modal` state applies to the application not to Windows.

> ### Before
>
> image

**Comment by gerhardol** [comment]
I believe that minimizing from a modal is good to have.

**Comment by RussKie** [comment]
To be honest, I do think this discussion is counter-productive. As @cristianst85 pointed earlier this dialog is inconsistent with the significant majority of our short-lived modal dialogs, and I agree with him.

First of all, as I mentioned earlier, we have many dialogs with only "close" button enabled. Good UX is a consistent UX.

Secondly, the minimise functionality is **completely** irrelevant for this dialog as it requires no user input (like ticket number) - it is invoked by selecting a specific commit in the revision grid.

Thirdly, enabling only the Minimize button "pollutes" the window's caption:
image
...and it **does not** minimise the app, only the dialog itself.

Lastly, I don't really buy into the argument that one must be able to minimize the app when creating a branch (using your example). ALT+TAB to the source of the data (e.g., a browser for a ticket number), or simply dismiss the dialog with ESC, minimise the app.

**Comment by mstv** [comment]
minimize_modal

**Comment by cristianst85** [comment]
@mstv, what you are seeing is a side effect. The window goes into foreground (not sure exactly what's the correct term), but does not minimize. Minimize your browser and try again to see what happens.

**Comment by RussKie** [comment]
https://github.com/gitextensions/gitextensions/assets/4403806/325996c5-ecfc-40ad-a139-926a353e6934

**Comment by RussKie** [comment]
> P.S. I am starting to feel bad for bringing this up in the first place. @RussKie, thank you for you hard work on this. ;)

All good. We're arguing because we're all feeling passionate about it. (It doesn't stop us from being pissed off in the interim 😆)

**Comment by RussKie** [comment]
I'm going to merge this in a day or so, so I can get on with a release of a new version.

The showing of the control box elements for short-lived dialogs needs to be assessed holistically. As has been shown by several people above the minimise button does not minimise the whole app, only the dialog itself. 
If you like to pursue this issue further, please open a new issue.

**Comment by mstv** [comment]
The `Checkout` button is not the default button any longer!
